### ISSUE TITLE
[Tracking Issue] Relax training M0 migration and polishment

### ISSUE THREAD

**Issue by SiriusNEO** [mlc-ai/relax#97]
In [this fork](https://github.com/ACMClass-TVM-20/relax-training) @Ubospica and me collaborated on the development of a training workflow on relax. We finished M0 and the efforts included:
- An automatic differentiation pass on relax high level operators.
- Some relax operators for training.
- Gradients registered for some relax operators.
- A framework supports training with different optimizers.
- A trainer wrapper.

Now these results are being migrated to new struct info infra. During this migration, we also try our best to polish our previous work. Related PRs and Progress:

- [x] (Dependency) Operator Legalizer
   - [x] Legalizer V0 https://github.com/mlc-ai/relax/pull/96
   - [x] Legalize more operators https://github.com/mlc-ai/relax/pull/101
- [x] (Dependency) TOPI `collapse_sum` patch https://github.com/mlc-ai/relax/pull/102
- [x] Some related operators
   - [x] `collapse_sum_to`, `collapse_sum_like` https://github.com/mlc-ai/relax/pull/87
   - [x] `cross_entropy`, `log_softmax`, `nll_loss` https://github.com/mlc-ai/relax/pull/94
   - [x] `exp`  https://github.com/mlc-ai/relax/pull/100
   - [x] `abs` (For L1Loss)  https://github.com/mlc-ai/relax/pull/113
- [x] Operator Gradients
  - [x] Gradients in new infra. Gradients for `log_softmax` and `nll_loss`, Gradients for `split` and `concat`. `collapse_sum` erasure. https://github.com/mlc-ai/relax/pull/98
  - [x] Gradients for nll_loss and conv2d https://github.com/mlc-ai/relax/pull/114
- [x] Automatic Differentiation Pass
  - [x] CopyWithNewParams utils https://github.com/mlc-ai/relax/pull/91
  - [x] AD Pass (Some detailed checkpoints can be found in this PR) https://github.com/mlc-ai/relax/pull/103
 - [x] Optimizer & Trainer APIs
   - [x] Optimizer https://github.com/mlc-ai/relax/pull/107
   - [x] Refactor on Optimizer and Gradient https://github.com/mlc-ai/relax/pull/121
   - [x] Loss functions https://github.com/mlc-ai/relax/pull/112
   - [x] Trainer https://github.com/mlc-ai/relax/pull/115

**Comment by Ubospica** [comment]
Glad to see the progress so far!
